keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,179 @@
+---
+kep-number: 0034
+title: Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles
+authors:
+  - "@Monkeyanator"
+editors:
+  - "@dashpole"
+owning-sig: sig-instrumentation
+reviewers:
+  - "@Random-Liu"
+  - "@bogdandrutu"
+approvers:
+  - "@brancz"
+creation-date: 2018-12-04
+last-updated: 2018-12-07
+status: provisional
+---
+
+# Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles
+
+## Table of Contents
+
+- [Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles](#leveraging-distributed-tracing-to-understand-kubernetes-object-lifecycles)
+    - [Table of Contents](#table-of-contents)
+    - [Summary](#summary)
+    - [Motivation](#motivation)
+        - [Goals](#goals)
+        - [Non-Goals](#non-goals)
+    - [Proposal](#proposal)
+        - [Architecture](#architecture)
+            - [Trace lifecycle](#trace-lifecycle)
+            - [Root spans](#root-spans)          
+            - [Context propagation](#context-propagation)          
+        - [Out-of-tree changes](#out-of-tree-changes)
+            - [Mutating admission webhook](#mutating-admission-webhook)
+            - [Tracing best-practices documentation](#tracing-best-practices-documentation)
+        - [In-tree changes](#in-tree-changes)
+            - [Trace utility package](#trace-utility-package)
+            - [Basic Kubernetes component trace instrumentation](#basic-kubernetes-component-trace-instrumentation)
+    - [Graduation Requirements](#graduation-requirements)
+    - [Implementation History](#implementation-history)
+
+## Summary
+
+This Kubernetes Enchancement Proposal (KEP) introduces a model for adding distributed tracing to Kubernetes object lifecycles. The inclusion of this trace instrumentation will mark a significant step in making Kubernetes processes more observable, understandable, and debuggable.
+
+
+## Motivation
+
+Debugging latency issues in Kubernetes is an involved process. There are existing tools which can be used to isolate these issues in Kubernetes, but these methods fall short for various reasons. For instance:
+
+* **Logs**: are fragmented, and finding out which process was the bottleneck involves digging through troves of unstructured text. In addition, logs do not offer higher-level insight into overall system behavior without an extensive background on the process of interest. 
+* **Events**: in Kubernetes are only kept for an hour by default, and don't integrate with visualization of analysis tools. To gain trace-like insights would require a large investment in custom tooling.
+* **Latency metrics**: are gathered in some places, but these don't provide understanding into _why_ a given process was slow.
```

Comment:
Part of the reason why latency metrics aren't a great way to determine _why_ a process was slow is for carnality reasons.  You wouldn't, for example, want to attach the container ID of a hypothetical `container_start_latency` metric because you would be making a new metric stream for each container, each with only a single sample taken.